### ISSUE TITLE
Rodeo Bokeh support 

### ISSUE THREAD

**Issue by raghavkgarg** [yhat/rodeo#439]
Bokeh Output does not show up in plot

**Comment by TakenPilot** [comment]
Ah, thanks for letting me know!

Could you give me an example bokeh script and what the output should look like?  I have a series of example scripts that I run before every release to determine if everything is working, and it would be useful to add a bokeh one.

**Comment by raghavkgarg** [comment]
If i am not wrong , I should expect plot to be shown in Plot window instead of opening HTML in browser by below code

from bokeh.io import output_notebook, show
output_notebook()
from bokeh.charts import Histogram
from bokeh.sampledata.iris import flowers as data
hist = Histogram(data, values="petal_length", color="species", legend="top_right", bins=12)
show(hist)

It is showing an error "Plot must be png, svg, html or javascript." in Plot area

**Comment by rgbkrk** [comment]
We have a bunch of React [packages in nteract](https://github.com/nteract) you can steal freely from/use/contribute to.

<img width="801" alt="screen shot 2016-08-29 at 9 18 32 am" src="https://cloud.githubusercontent.com/assets/836375/18058518/9b4661ac-6dc9-11e6-83ba-475775d23031.png">

We handle the display area in a similar way to the notebook using [react-jupyter-display-area](https://github.com/nteract/react-jupyter-display-area), which under the covers uses [transformime-react](https://github.com/nteract/transformime-react) (mimetype from jupyter message protocol --> React Element).

Generally speaking though, bokeh needs a global window context to work with both the initially loaded Bokeh script and the plots down below.

**Comment by TakenPilot** [comment]
Wow, those are really neat!  You're much better at React than I am.  : )

Right now, Rodeo is assuming that each item that IPython returns is self-contained and therefore can be exported and/or displayed as a self-contained item.  This works with most plots because they're either a single html page that can be put in an iframe, or an image.

That bokeh script is returning two objects from IPython: a `text/html` and a `application/javascript`.  It's expecting them both to be put in the same context as you said, and that isn't the case.

Any ideas about situations where we can assume two objects from IPython should be put into the same iframe?  I don't want to put anything into a general context, because that's where Rodeo runs.  Also, iframes in HTML5 can be sandboxed.

To see what I'm talking about, check out the output of your script in the DevTools:  

<img width="814" alt="screen shot 2016-08-30 at 8 36 22 am" src="https://cloud.githubusercontent.com/assets/4154804/18089231/de2cfaee-6e8c-11e6-961d-304ee9de8c97.png">

Rodeo detects the html page, and rightly intercepts it and put the html into a temporary file to be served from the file system in an iframe.  This is big advantage for memory and security (and just not crashing).  The javascript isn't something that I've considered, so it isn't intercepted, but Rodeo also doesn't know how to display it.

Any ideas?

**Comment by rgbkrk** [comment]
You could make it behave in a notebook like way by hiding old HTML outputs in the same iframe, while making them hidden. The main thing they're looking for is for their JS library to be loaded for a later context.

**Comment by TakenPilot** [comment]
True.  I worry about trying to do anything in a notebook-like way, since notebook already exists and works just fine for what it is for.  For the workflow in Rodeo, it's nice to have each execution be distinct so you can browse around all the plots that have been created as we play around with the data.

I looked at what Bokeh is outputting for html, and it's just an empty html page that the later JavaScript is filling in.  It's weird that Bokeh would separate them into two pieces like that instead of just putting the JS into the HTML and outputting a single page that could be served.  It makes it difficult to be generic and safe when they're expecting us to know how to combine the two pieces together.

**Comment by TakenPilot** [comment]
Playing with the library further, Bokeh seems _really_ notebook specific.  It's expecting cells to exist and for them to be query-able.  In Rodeo, the concept of cells doesn't really exist in the same way.

There might be a markdown-like format later (I'm about 3/4 done the implementation of this), but Rodeo really shines as a way to explore and manipulate data within a workspace designed for that.  Notebook is more about overriding previous work and a whole lot of scrolling.  😉 

Maybe I should make a PR so that Bokeh can output to IPython without the assumption of overriding the old output.

**Comment by rgbkrk** [comment]
Totally. I'm not too pleased with the backend code expecting a certain frontend code response (having requirejs defined for example, as well as specific versions of jQuery, lodash, widgets, etc.). The way we're approaching deeper and simpler viz is with custom mimetypes: https://github.com/nteract/nteract/pull/661

JupyterLab is the same, though coding it for ipywidgets. I picked victory there to make it simple.

**Comment by rgbkrk** [comment]
Here's a WIP for Plotly.js:

https://github.com/nteract/nteract/pull/662

Relevant commit is https://github.com/nteract/nteract/commit/60e069beecee20655bb109476a7df29b8384043c

**Comment by rgbkrk** [comment]
Ping @birdsarah!
